ReSharper.FSharp/src/FSharp/FSharp.Psi/src/Util/FcsTypeMappingUtil.cs

```diff
@@ -333,6 +333,15 @@ private static FSharpCheckFileResults GetCheckResults(this IFSharpTreeNode fsTre
     public static FSharpType TryGetFcsType([NotNull] this IFSharpTreeNode treeNode) =>
       TryGetFcsType(treeNode, treeNode.GetDocumentRange());
 
+    [CanBeNull]
+    public static FSharpType TryGetFcsType([NotNull] this IFSharpPattern pattern) =>
+      pattern switch
+      {
+        IReferencePat pat => (pat.GetFcsSymbol() as FSharpMemberOrFunctionOrValue)?.FullType,
+        IOptionalValPat { Pattern: IReferencePat innerPattern } => innerPattern.TryGetFcsType()?.GenericArguments[0],
+        { } => (pattern as IFSharpTreeNode).TryGetFcsType()
```

Comment:
`pattern` is annotated as `NotNull`. Why is the `{ }` required?

`IFSharpPattern` implements `IFSharpTreeNode`. Why is the `as` cast required?

Reply:
> Why is the { } required?

During code editing, for some reason, _ pattern didn't work, and I used a different one (hm, but now the first one works).

> IFSharpPattern implements IFSharpTreeNode. Why is the as cast required?

To call an existing extension with logic for IFSharpTreeNode instead of an infinite recursive call.